src/main/java/edu/harvard/iq/dataverse/api/Access.java

```diff
@@ -1787,37 +1761,41 @@ private boolean isAccessAuthorized(DataFile df, String apiToken) {
             }
         }
         
-        if (!restricted && !embargoed) {
-            // And if they are not published, they can still be downloaded, if the user
+        
+
+        //The one case where we don't need to check permissions
+        if (!restricted && !embargoed && published) {
+            // If they are not published, they can still be downloaded, if the user
             // has the permission to view unpublished versions! (this case will 
             // be handled below)
-            if (published) { 
-                return true;
-            }
+            return true;
         }
         
+        //For permissions check decide if we havce a session user, or an API user
         User user = null;
         
         /** 
          * Authentication/authorization:
-         * 
-         * note that the fragment below - that retrieves the session object
-         * and tries to find the user associated with the session - is really
-         * for logging/debugging purposes only; for practical purposes, it 
-         * would be enough to just call "permissionService.on(df).has(Permission.DownloadFile)"
-         * and the method does just that, tries to authorize for the user in 
-         * the current session (or guest user, if no session user is available):
          */
         
-        if (session != null) {
+        User apiTokenUser = null;
```

Comment:
The logic used to handle 3 cases separately, mostly so that fine logging could be done to indicate which branch was being used: session - logged in user, guest user, and apitoken user. One tricky thing is that both session and apitoken calls could return a GuestUser and this wasn't broken out.

Reply:
I like how the code has been reorganized and simplified. And the logic seems correct. But could you please clarify the "this wasn't broken out" part above. Aside from the old code being convoluted and unwieldy, was there actually a problem with the old logic, was there a situation where the guest user returned by the apitoken call would overwrite an authenticated user from the session? - I've stared at the old code for a while now and I'm not seeing that, but figured I would ask to confirm.

(We may have assumed in the past that nobody would be calling the api with _both_ a session and a token in real life).

Reply:
I don't think it was broken, so no real issue. I think my comment related to the old line 1877 and the code after it - there was no distinction between guest coming back from a session or from the apitoken logic. In refactoring, I had to make sure that both were still handled, e.g. in line 1790 below.

Reply:
I was looking at that code after line 1877 - it was just meaningless, wasn't it? I.e., it doesn't seem to be doing anything that the line 1861 hasn't already attempted. 
It just looks like the old code may have been written with the assumption that `session == null` when it's a direct API call - and it should never be null of course.